src/parsers/manifest/dash/wasm-parser/rs/processor/attributes.rs

```diff
@@ -0,0 +1,393 @@
+use crate::errors::ParsingError;
+use crate::events::AttributeName::*;
+
+pub fn report_mpd_attrs(e : &quick_xml::events::BytesStart) {
+    for res_attr in e.attributes() {
+        match res_attr {
+            Ok(attr) => match attr.key {
+                b"id" => Id.try_report_as_string(&attr),
+                b"profiles" => Profiles.try_report_as_string(&attr),
+                b"type" => Type.try_report_as_string(&attr),
+                b"availabilityStartTime" => AvailabilityStartTime.try_report_as_string(&attr),
+                b"availabilityEndTime" => AvailabilityEndTime.try_report_as_string(&attr),
+                b"publishTime" => PublishTime.try_report_as_string(&attr),
+                b"mediaPresentationDuration" =>
+                    MediaPresentationDuration.try_report_as_iso_8601_duration(&attr),
+                b"minimumUpdatePeriod" =>
+                    MinimumUpdatePeriod.try_report_as_iso_8601_duration(&attr),
+                b"minBufferTime" =>
+                    MinBufferTime.try_report_as_iso_8601_duration(&attr),
+                b"timeShiftBufferDepth" =>
+                    TimeShiftBufferDepth.try_report_as_iso_8601_duration(&attr),
+                b"suggestedPresentationDelay" =>
+                    SuggestedPresentationDelay.try_report_as_iso_8601_duration(&attr),
+                b"maxSegmentDuration" =>
+                    MaxSegmentDuration.try_report_as_iso_8601_duration(&attr),
+                b"maxSubsegmentDuration" =>
+                    MaxSubsegmentDuration.try_report_as_iso_8601_duration(&attr),
+                _ => {},
```

Comment:
Here, what does happen if there is no 'key' property ? Does it match with this _ => {}, or is it even possible that it does not include the 'key' attribute ?

Reply:
> what does happen if there is no 'key' property

Rust doesn't have the same model than JS/TS where an object property could be defined or undefined.

Also `match` expressions always need to be exhaustive to compile, so anything that is not catched there goes into the last, noop branch.

For when some property can be optional, Rust idiomatic way of dealing with it is closer to Scala/Haskell where you define it through its type, as an Option<T>/Maybe<T>.
Rust has [`Option<T>`](https://doc.rust-lang.org/std/option/) and to Handle these, you generally resort to pattern-matching.

---

For this code here you could install a Rust language server in your text editor and go to the definition of `key` (https://github.com/tafia/quick-xml/blob/4855faee830cd5e9100429567cb1ac587f2db2d7/src/events/attributes.rs#L84) and find out it is a:

> `&'a [u8]`

Which is a reference (the `&` part) to a "slice" (some sort of array) of bytes (`[u8]`).
The `'a` part refers to this reference's lifetime*.

So it can never be undefined, it is always some reference to some bytes, and we're matching it against different ascii combinations (by using the `b""` syntax), fallbacking to nothing if nothing match (the `_ => {}` part).  

\* "Lifetime" and "References" are a little complicated to explain briefly, so I added a wall of text below to try my best to explain what it is and what `'a` is.

---

Rust has a concept of "ownership" that is associated to each variable declared. The idea is basically that when the variable goes out of the scope of its owner, its linked resources are directly liberated.

By default. there is only one owner of a variable (it's actually more complicated, because you can use things like https://doc.rust-lang.org/std/rc/index.html but that's just complicating things). If a function just gives the variable as is as an argument to another function, it loses its ownership and as such cannot use that variable anymore.

A way around that is using references, by adding a `&` before it.
Technically, they generally are just a pointer to the underlying data (sometimes not only, for example a `&[u8]` like here also includes information about the slice's length) but that's internal things to Rust and we're not supposed to rely on that.

You can use references much more freely than the original values (though there are also rules depending on if you "borrow a reference" mutably or immutably but I would drift too far there).

The issue with reference though, is that as they aren't the owner of the data, it depends on when the underlying value will be freed (when it goes out of scope of its owner).

In our case, we have just some `struct`, `Attribute` storing directly a reference without indicating who is the owner of that data. So the compiler is a little lost, it cannot prove easily that the variable won't be used after the original value is freed. 
 
That's where the concept of lifetimes comes in. With it, we can link the lifetime of that reference to e.g. another reference whose owner is known at compile-time.

Showing where it is done for `Attribute` (singular form) is not the easiest one but e.g. for `Attributes` (plural), we can see it can be linked to the `buf` argument given to its `new` static function: https://github.com/tafia/quick-xml/blob/4855faee830cd5e9100429567cb1ac587f2db2d7/src/events/attributes.rs#L36 

More information about ownership here: https://doc.rust-lang.org/book/ch04-01-what-is-ownership.html (it's one of the most complicated aspect when starting Rust, so don't be afraid if you don't get it right away!).

Reply:
Done